### ISSUE TITLE
BUG: Complex indexing changes axis order

### ISSUE THREAD

**Issue by francois-rozet** [numpy/numpy#29266]
### Describe the issue:

Unexpected axis order with mildly complex indexing. I don't think this is the expected behavior for most users.

### Reproduce the code example:

```python
>>> import numpy as np
>>> x = np.random.randn(3, 5, 7, 11, 13, 17)
>>> x[:, 0, np.array([0, 1]), :].shape  # expected shape
(3, 2, 11, 13, 17)
>>> x[:, 0, np.array([0, 1]), :, 0].shape  # two first axes are flipped
(2, 3, 11, 17)
```

### Python and NumPy Versions:

```
>>> import sys, numpy; print(numpy.__version__); print(sys.version)
2.3.1
3.12.5 (main, Aug 14 2024, 05:08:31) [Clang 18.1.8 ]
```

### Context for the issue:

With `torch`, the same operation results in the expected shape:

```python
>>> import torch
>>> x = torch.randn(3, 5, 7, 11, 13, 17)
>>> x[:, 0, torch.tensor([0, 1]), :, 0].shape
torch.Size([3, 2, 11, 17])
```

**Comment by francois-rozet** [comment]
@seberg I think this is a critical bug.

**Comment by seberg** [comment]
It's not a bug, it is documented behavior.

**Comment by francois-rozet** [comment]
Oh. Can you point me toward the documentation of this behavior?

**Comment by rkern** [comment]
https://numpy.org/devdocs/user/basics.indexing.html#combining-advanced-and-basic-indexing
